beacon_node/eth2-libp2p/src/rpc/handler.rs

```diff
@@ -461,46 +458,32 @@ where
-        if let Some((request_id, err)) = self.pending_error.pop() {
+        if let Some((request, err)) = self.pending_error.pop() {
             // Returning an error here will result in dropping the peer.
-            match err {
-                ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Apply(
-                    RPCError::InvalidProtocol(protocol_string),
-                )) => {
-                    // Peer does not support the protocol.
-                    // TODO: We currently will not drop the peer, for maximal compatibility with
-                    // other clients testing their software. In the future, we will need to decide
-                    // which protocols are a bare minimum to support before kicking the peer.
-                    error!(self.log, "Peer doesn't support the RPC protocol"; "protocol" => protocol_string);
-                    return Ok(Async::Ready(ProtocolsHandlerEvent::Custom(
-                        RPCEvent::Error(request_id, RPCError::InvalidProtocol(protocol_string)),
-                    )));
-                }
-                ProtocolsHandlerUpgrErr::Timeout | ProtocolsHandlerUpgrErr::Timer => {
-                    // negotiation timeout, mark the request as failed
-                    debug!(self.log, "Active substreams before timeout"; "len" => self.outbound_substreams.len());
-                    return Ok(Async::Ready(ProtocolsHandlerEvent::Custom(
-                        RPCEvent::Error(
-                            request_id,
-                            RPCError::Custom("Protocol negotiation timeout".into()),
-                        ),
-                    )));
-                }
-                ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Apply(err)) => {
-                    // IO/Decode/Custom Error, report to the application
-                    debug!(self.log, "Upgrade Error"; "error" => format!("{}",err));
-                    return Ok(Async::Ready(ProtocolsHandlerEvent::Custom(
-                        RPCEvent::Error(request_id, err),
-                    )));
-                }
-                ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(err)) => {
-                    // Error during negotiation
-                    debug!(self.log, "Upgrade Error"; "error" => format!("{}",err));
-                    return Ok(Async::Ready(ProtocolsHandlerEvent::Custom(
-                        RPCEvent::Error(request_id, RPCError::Custom(format!("{}", err))),
-                    )));
-                }
-            }
+            let outcome = match err {
+                ProtocolsHandlerUpgrErr::Timer => RPCError::InternalError("Timer failed"),
+                ProtocolsHandlerUpgrErr::Timeout => RPCError::StreamTimeout,
+                ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)) => e,
+                ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                    NegotiationError::Failed,
+                )) => RPCError::UnsuportedProtocol,
```

Comment:
It will be very useful to know which type of protocol is unsupported here. We can determine this from the `request` type